### PR TITLE
Add randomness doc

### PR DESCRIPTION
**Pull request by chasefleming**

### CODE REVIEW

File: docs/build/advanced-concepts/randomness.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+- **Gaming:** Integrates fairness and unpredictability into gameplay, enhancing user engagement without delays.
+- **NFTs:** Facilitates the creation of uniquely randomized traits in NFTs quickly, adding to their rarity and value.
+- **Lotteries & Draws:** Offers instant and verifiably fair random selection for lotteries, solidifying trust in real-time.
+- **DeFi Protocols:** Enables rapid and innovative random reward systems within decentralized finance.
+- **DAOs:** Assists in unbiased voting and task assignments through immediate randomness.
+- **Broad Applications:** Extends to any domain requiring impartial randomization, from asset distribution to security mechanisms, all with the added benefit of on-demand availability.
+
+## History - Distributed Random Beacon
+
+Within the Flow protocol, the heart of randomness generation lies in the "distributed random beacon." This module generates randomness that is distributed across the network while adhering to established safety and security standards. The output from the randomness beacon is a sequence of random bytes that are unpredictable and impartial.
```
Comment (tarakby) [inline docs/build/advanced-concepts/randomness.md:23]: Alex's suggestion from Notion
```suggestion
Within the Flow protocol, the heart of randomness generation lies in the "distributed random beacon." This module generates randomness that is distributed across the network while adhering to established cryptographic and security standards. The output from the randomness beacon is a sequence of random bytes that are unpredictable and impartial.
```

File: docs/build/advanced-concepts/randomness.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+For over three years, the beacon has ensured protocol security by assigning verification nodes to oversee block computations. For those interested in a more detailed exploration of the randomness beacon and its inner workings, you can read [the technical deep dive on the Flow forum](https://forum.flow.com/t/secure-random-number-generator-for-flow-s-smart-contracts/5110).
```
Comment (tarakby) [inline docs/build/advanced-concepts/randomness.md:25]: Alex's suggestion from Notion:
```suggestion
For over three years, the beacon has ensured protocol security by selecting which consensus node gets to propose the next block and assigning verification nodes to oversee block computations. For those interested in a more detailed exploration of the randomness beacon and its inner workings, you can read [the technical deep dive on the Flow forum](https://forum.flow.com/t/secure-random-number-generator-for-flow-s-smart-contracts/5110).
```

File: docs/build/advanced-concepts/randomness.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+🚨 A transaction can atomically revert all its action during its runtime and abort. Therefore, it is possible for a transaction calling into your smart contract to post-select favorable results and revert the transaction for unfavorable results.
+</Callout>
+
+In other words, transactions submitted by a non-trusted party are able to reject their results after the random is revealed.
+
+<Callout type="info">
+💡 **Post-selection** - the ability for transactions to reject results they don't like - is inherent to any smart contract platform that allows transactions to roll back atomically. See this very similar [Ethereum example](https://consensys.github.io/smart-contract-best-practices/development-recommendations/general/public-data/).
+</Callout>
+
+The central aspect that you as a developer have to think about is the following scenario:
```
Comment (tarakby) [inline docs/build/advanced-concepts/randomness.md:60]: minor suggestion but feel free to ignore
```suggestion
The central aspect that a contract developer needs to think about is the following scenario:
```

File: docs/build/advanced-concepts/randomness.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+- **DeFi Protocols:** Enables rapid and innovative random reward systems within decentralized finance.
+- **DAOs:** Assists in unbiased voting and task assignments through immediate randomness.
+- **Broad Applications:** Extends to any domain requiring impartial randomization, from asset distribution to security mechanisms, all with the added benefit of on-demand availability.
+
+## History - Distributed Random Beacon
+
+Within the Flow protocol, the heart of randomness generation lies in the "distributed random beacon." This module generates randomness that is distributed across the network while adhering to established safety and security standards. The output from the randomness beacon is a sequence of random bytes that are unpredictable and impartial.
+
+For over three years, the beacon has ensured protocol security by assigning verification nodes to oversee block computations. For those interested in a more detailed exploration of the randomness beacon and its inner workings, you can read [the technical deep dive on the Flow forum](https://forum.flow.com/t/secure-random-number-generator-for-flow-s-smart-contracts/5110).
+
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:26]: short section on history. Might need @tarakby's revisions 

```suggestion

### The History and Limitations of `unsafeRandom` (now deprecated)

Cadence has historically provided the `unsafeRandom` function to return a pseudo-random number. The stream of random numbers produced was potentially unsafe in the following two regards:

1. The sequence of random numbers is potentially predictable by transactions within the same block and by other smart contracts calling into your smart contract.
2. A transaction calling into your smart contract can potentially bias the sequence of random numbers which your smart contract internally generates. Currently, the block hash seeds `unsafeRandom`. Consensus nodes can *easily* bias the block hash and **influence the seed for `unsafeRandom`**.

⚠️ Note `unsafeRandom` will be deprecated in the next Cadence release. **Developers are advised to refrain from using `unsafeRandom.`**
```

File: docs/build/advanced-concepts/randomness.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+Although Cadence exposes safe randomness generated by the Flow protocol via `revertibleRandom`, there is an additional safety-relevant aspect that developers need to be mindful about.
+
+The `revertibleRandom` function can be used safely in some applications where the transaction results are NOT deliberately reverted after the random number is revealed (i.e. a contract distributing random NFTs to registered users or onchain lucky draw). However, if applications require a non-trusted party (for instance app users) to submit a transaction calling a randomized (non-deterministic) contract, the stream of random numbers produced is potentially unsafe in the following two regards:
+
+1. The sequence of random numbers is potentially predictable by transactions within the same block and by other smart contracts calling into your smart contract.
+2. A transaction calling into your smart contract can potentially bias the sequence of random numbers which your smart contract internally generates. Currently, the block hash seeds `revertibleRandom`. Consensus nodes can *easily* bias the block hash and **influence the seed for** `revertibleRandom`.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:48]:
```suggestion
The `revertibleRandom` function can be used safely in some applications
where the transaction results are _not_ deliberately reverted after the random number is revealed
(i.e. a contract distributing random NFTs to registered users or onchain lucky draw).
However, if applications require a non-trusted party (for instance app users) to submit a transaction
calling a randomized (non-deterministic) contract, the developer must explicitly protect the stream of
random numbers to not break the security guarantees:
```

File: docs/build/advanced-concepts/randomness.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+<Callout type="warning">
+🚨 A transaction can atomically revert all its action during its runtime and abort. Therefore, it is possible for a transaction calling into your smart contract to post-select favorable results and revert the transaction for unfavorable results.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:51]:
```suggestion
🚨 A transaction can atomically revert all its action during its runtime and abort.
Therefore, it is possible for a transaction calling into your smart contract to post-select
favorable results and revert the transaction for unfavorable results.
```

File: docs/build/advanced-concepts/randomness.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+In simpler terms, the native secure randomness provided by the protocol can now be safely utilized within Cadence smart contracts and is available to all developers on Flow and the FVM.
+
+While the source values are safely generated by the Random Beacon and transmitted into the execution environment via the committing transaction, using the raw values from this contract does not guarantee non-revertible randomness. In addition to the commit-reveal pattern described below, it is also recommended to use the source values with a pseudo-random number generator (PRG) to generate an arbitrarily-long sequence of random values.
+
+Adding a safe pattern to reveal randomness without the possibility of conditional transaction reversion unlocks applications relying on randomness. By providing examples of commit-reveal implementations we hope to foster a more secure ecosystem of decentralized applications and encourage developers to build with best practices.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:82]:
```suggestion
The recommended way to mitigate the problems above is via a commit-reveal scheme.
The scheme involves two steps: commit and reveal. During the commit phase, the user
transaction commits to accepting the future output of a smart contract where the last
remaining input is an unknown stream of random numbers. The smart contract stores this
commitment on the blockchain. At the current level of optimization, the reveal phase can
start as early as the next block, when the beacon's source of randomness becomes available.
With a second transaction, the smart contract can be executed to explicitly generate the
output now that the stream of random numbers can be seeded (plus suitable salt).

There are ideas how to optimize the developer experience further in the future.
For example, a transaction could delegate part of its gas to an independent transaction
it spawns. Such solution would conceptually also be a commit-and-reveal scheme, just immediately
happening within the same block. Until we eventually get to this next level, developers may need
to implement their own commit-reveal. In Cadence, it is clean and short.

Protocol improvements (documented in [FLIP 120](https://github.com/onflow/flips/blob/main/cadence/20230713-random-function.md))
expose the random beacon to the FVM and Cadence where it can be used to seed pseudo-random number generators [PRNGs] for smart contracts.

While both are backed by Flow's Random Beacon it is important for developers to mindfully choose between `revertibleRandom` or
seeding their own PRNG utilizing the `RandomBeaconHistory` smart contract:

- Under the hood, the FVM also just instantiates a PRNG for each transaction that `revertibleRandom` draws from.
  Though, with `revertibleRandom` a developer is calling the PRNG that is controlled by the transaction,
  which also has the power to abort and revert if it doesn't like `revertibleRandom`'s outputs.
  `revertibleRandom` is only suitable for smart contract functions that exclusively run within the trusted transactions.
- In contrast, using the `RandomBeaconHistory` means to use a deterministically-seeded PRNG.
  The `RandomBeaconHistory` is key for effectively implementing a commit-and-reveal scheme.
  During the commit phase, the user commits to proceed with a future source of randomness,
  which is revealed after the commit transaction concluded.
  For each block, the `RandomBeaconHistory` automatically stores the subsequently generated source of randomness.

Adding a safe pattern to reveal randomness without the possibility of conditional transaction reversion unlocks applications relying on randomness.
By providing examples of commit-reveal implementations we hope to foster a more secure ecosystem of decentralized applications
and encourage developers to build with best practices.

In simpler terms, the native secure randomness provided by the protocol can now be safely utilized within Cadence smart contracts
and is available to all developers on Flow and the FVM.
```

File: docs/build/advanced-concepts/randomness.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+## FLIP 123
+
+On Flow, we have absorbed all security complexity into the platform.
+
+[FLIP 123: On-chain Random beacon history for commit-reveal schemes](https://github.com/onflow/flips/blob/main/protocol/20230728-commit-reveal.md#flip-123-on-chain-random-beacon-history-for-commit-reveal-schemes) was introduced to provide a safe pattern to use randomness in transactions so that it's not possible to revert non-favorable randomized transaction results.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:88]:
```suggestion
[FLIP 123: On-chain Random beacon history for commit-reveal schemes](https://github.com/onflow/flips/blob/main/protocol/20230728-commit-reveal.md#flip-123-on-chain-random-beacon-history-for-commit-reveal-schemes) was introduced to provide a safe pattern to use randomness in transactions so that it's not possible to revert unfavorable randomized transaction results.
```

File: docs/build/advanced-concepts/randomness.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+<Callout type="info">
+💡 While the commit-and-reveal scheme mitigates post-selection of results by adversarial clients, Flow’s secure randomness additionally protects against any pre-selection vulnerabilities (like biasing attacks by byzantine miners).
+</Callout>
+
+A commit-reveal scheme can be implemented as follows. The coin toss example described earlier will be used for illustration:
+
+- When a user submits a bidding transaction, the bid amount is transferred to the coin toss contract, and the block height where the bid was made is stored. This is a commitment by the user to use the SoR at the current block. Note that the current block's `SoR_A` isn't known to the transaction execution environment, and therefore the transaction has no way to inspect the random outcome and predict the coin toss result. The current block's `SoR_A` is only available once added to the history core-contract, which only happens at the end of the block's execution. The user may also commit to using an SoR of a future block, which is still unknown at the time the bid is made.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:97]:
```suggestion
- When a user submits a bidding transaction, the bid amount is transferred to the coin toss contract, and the block height where the bid was made is stored. This is a commitment by the user to use the SoR at the current block. Note that the current block's `SoR_A` isn't known to the transaction execution environment, and therefore the transaction has no way to inspect the random outcome and predict the coin toss result. The current block's `SoR_A` is only available once added to the history core-contract, which only happens at the end of the block's execution. The user may also commit to using an SoR of some future block, which is equally unknown at the time the bid is made.
```

File: docs/build/advanced-concepts/randomness.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+- The coin toss contract may grant the user a limited window of time (i.e a block height range) to reveal the results and claim any winnings. Failing to do so, the bid amount remains in the coin toss contract.
+- The user can submit a second transaction to call the coin toss contract and resolve the bid. The coin toss contract looks up the committed block height of the user and checks it has already passed. The contract uses the block height to query the past-block's `SoR_A` on the core-contract `RandomBeaconHistory`.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:99]:
```suggestion
- The coin toss contract may grant the user a limited window of time (i.e a block height range) to send a second transaction for resolving the results and claim any winnings. Failing to do so, the bid amount remains in the coin toss contract.

  Within that reveal transaction, the user calls the coin toss contract, looks us up the block height at which the block was committed and checks that it has already passed. The contract queries that block's `SoR_A` from the core-contract `RandomBeaconHistory` via block height.
```

File: docs/build/advanced-concepts/randomness.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+- The coin toss contract uses a PRG seeded by the queried `SoR_A` and diversified using a specific information to the use-case (a user ID or resource ID for instance). Diversification does not add new entropy, but it avoids generating the same outcome for different use-cases. If a diversifier (or salt) isn't used, all users that committed a bid on the same block would either win or lose.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:100]:
```suggestion
- The coin toss contract uses a PRG seeded with the queried `SoR_A` and diversified using a specific information to the use-case (a user ID or resource ID for instance). Diversification does not add new entropy, but it avoids generating the same outcome for different use-cases. If a diversifier (or salt) isn't used, all users that committed a bid on the same block would either win or lose.
```

File: docs/build/advanced-concepts/randomness.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+- The PRG is used to generate the random result and resolve the bid. Note that the user can make the transaction abort after inspecting a losing result. However, the bid amount would be lost anyway when the allocated window expires.
+
+The following lines of code illustrate a random coin toss, that cannot be gamed or biased. The reveal-and-commit scheme prevent clients from post-selecting favorable outcomes. 
+
+It proceeds with two phases: first commit using the hash of the values and in a later phase revealing the values.
```
Comment (AlexHentschel) [inline docs/build/advanced-concepts/randomness.md:105]:
```suggestion
The following lines of code illustrate a random coin toss that cannot be gamed or biased. The reveal-and-commit scheme prevent clients from post-selecting favorable outcomes.
```

**Review by AlexHentschel — COMMENTED** [review]
Porting my edits from [notion](https://www.notion.so/dapperlabs/Add-Page-to-Advance-Concepts-introducing-onchain-randomness-96000cc737d647198f1cf054e561e9e9?pvs=4). There were some aspects that were starting to become incorrect due to the brevity. We could either move those details into a separate tutorial. But if we can afford a bit more detail, I extended the text for correct presentation. Still tried to keep it very concise ... not sure where a good cutoff is. Please let me know your thoughts. Thanks.

File: docs/build/advanced-concepts/randomness.md (hunk 4 above)
Comment (tarakby) [reply docs/build/advanced-concepts/randomness.md:26]: There are two other less important aspects but we don't necessarily need to mention them:
 3. the PRNG used did not provide high statistical quality.
 4. The sequence of random numbers is the same for each transaction of the same block
 
 I'm happy with the current state if we want to keep this brief.

File: docs/build/advanced-concepts/randomness.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+Within the Flow protocol, the heart of randomness generation lies in the "distributed random beacon." This module generates randomness that is distributed across the network while adhering to established safety and security standards. The output from the randomness beacon is a sequence of random bytes that are unpredictable and impartial.
+
+For over three years, the beacon has ensured protocol security by assigning verification nodes to oversee block computations. For those interested in a more detailed exploration of the randomness beacon and its inner workings, you can read [the technical deep dive on the Flow forum](https://forum.flow.com/t/secure-random-number-generator-for-flow-s-smart-contracts/5110).
+
+## Guidelines for Safe Usage
+
+For usage of randomness where result abortion is not an issue, it is recommended to use the Cadence built-in function `revertibleRandom.` `revertibleRandom` returns a pseudo-random number and is also based on the safe Random Beacon.
+
+```cadence
+// Language reference:
```
Comment (sisyphusSmiling) [inline docs/build/advanced-concepts/randomness.md:32]: Could also be nice to link to the ecdao snippet: https://academy.ecdao.org/en/snippets/cadence-random

File: docs/build/advanced-concepts/randomness.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+- The coin toss contract uses a PRG seeded by the queried `SoR_A` and diversified using a specific information to the use-case (a user ID or resource ID for instance). Diversification does not add new entropy, but it avoids generating the same outcome for different use-cases. If a diversifier (or salt) isn't used, all users that committed a bid on the same block would either win or lose.
+- The PRG is used to generate the random result and resolve the bid. Note that the user can make the transaction abort after inspecting a losing result. However, the bid amount would be lost anyway when the allocated window expires.
+
+The following lines of code illustrate a random coin toss, that cannot be gamed or biased. The reveal-and-commit scheme prevent clients from post-selecting favorable outcomes. 
+
+It proceeds with two phases: first commit using the hash of the values and in a later phase revealing the values.
+
+```cadence
+/// --- Commit ---
```
Comment (sisyphusSmiling) [inline docs/build/advanced-concepts/randomness.md:108]:
```suggestion
// The code below is taken from the example CoinToss contract found in this project repo
// Source: https://github.com/onflow/random-coin-toss 

/// --- Commit ---
```

File: docs/build/advanced-concepts/randomness.md [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+
+For over three years, the beacon has ensured protocol security by assigning verification nodes to oversee block computations. For those interested in a more detailed exploration of the randomness beacon and its inner workings, you can read [the technical deep dive on the Flow forum](https://forum.flow.com/t/secure-random-number-generator-for-flow-s-smart-contracts/5110).
+
+## Guidelines for Safe Usage
+
+For usage of randomness where result abortion is not an issue, it is recommended to use the Cadence built-in function `revertibleRandom.` `revertibleRandom` returns a pseudo-random number and is also based on the safe Random Beacon.
+
+```cadence
+// Language reference:
+// https://developers.flow.com/cadence/language/built-in-functions#revertiblerandom
```
Comment (sisyphusSmiling) [inline docs/build/advanced-concepts/randomness.md:33]: Could also be nice to link to ecdao snippet

```suggestion
// https://developers.flow.com/cadence/language/built-in-functions#revertiblerandom
// Run the snippet here: https://academy.ecdao.org/en/snippets/cadence-random
```

**Review by gregsantos — APPROVED** [review]
